pkgs/build-support/kernel/closure.rs

```diff
@@ -0,0 +1,184 @@
+        out_dir: Path::new(&args[2]).to_path_buf(),
+        firmware_dir: Path::new(&firmware).to_path_buf(),
+        allow_missing: allow_missing,
+    };
+    let roots = (args[3..]).iter().cloned().filter_map(|root| {
+        let res = data.mod_resolve(&root);
+        if res.is_none(){
+            data.mod_missing(&root);
+        }
+        res
+    }).collect();
+
+    let cl = closure( |module| data.mod_dep(&module), &roots );
+
+    for module in cl {
+        data.copy_mod(&module);
+        data.copy_firmware(&module);
+    }
+}
+
+pub enum ModPath {
+    Path(PathBuf),
+    Builtin
+}
+
+struct ModulesData {
+    kernel: PathBuf,
+    out_dir : PathBuf,
+    firmware_dir: PathBuf,
+    allow_missing : bool
+}
+
+impl ModulesData {
+    fn mod_missing(&self, module : &OsString) {
+        eprintln!("modinfo failed!");
+        eprint!("that likely means that the module: ");
+        eprint_os(&module);
+        eprint!("\nwas not found in: ");
+        eprint_os(&self.kernel.clone().into_os_string());
+        if !self.allow_missing {
+            exit(1)
+        }
+    }
+    fn modinfo<P : AsRef<OsStr>>(&self, module : &OsString, field : P)
+        -> Option<Vec<u8>> {
+        let out = Command::new("modinfo")
+            .arg("-b").arg(&self.kernel)
+            .arg("-F").arg(field)
+            .arg(module)
+            .output()
+            .expect("failed to execute modinfo");
+        if out.status.success() {
+            Some(out.stdout)
+        } else {
+            self.mod_missing(&module);
+            None
+        }
+    }
+
+    fn mod_path(&self, module : &OsString) -> Option<ModPath> {
+        let fname = self.modinfo(module,"filename")?;
+        let fname = OsStr::from_bytes(&fname);
+        Some(
+            if fname == OsString::from("(builtin)") {
+                ModPath::Builtin
+            } else {
+                let path = Path::new(&fname);
+                ModPath::Path(path.to_path_buf())
+            }
+        )
+    }
+
+    fn copy_mod(&self, module : &OsString){
+        if let Some(modpath) = self.mod_path(&module) {
+            match modpath {
+                ModPath::Path(source) => {
+                    let target = self.out_dir.join(
+                        source.strip_prefix(&self.kernel).unwrap()
+                        );
+                    let target_dir = target.parent().unwrap();
+                    fs::create_dir_all(&target_dir).unwrap();
+                    fs::copy(&source,&target).unwrap();
+                }
+                ModPath::Builtin => { },
+            }
+        }
+    }
+
+    fn copy_firmware(&self, module : &OsString) {
+        if let Some(out) = self.modinfo(&module,"firmware") {
+            let firmware : Vec<OsString> = out
+                .split(|c| *c == b'\n')
+                .filter(|bs| !bs.starts_with(b"name:"))
+                .map(|bs| OsString::from_vec(Vec::from(bs)) )
+                .collect();
```

Comment:
The output format of `modinfo` should be described a bit more, maybe even parsed directly into a small data structure in `self.modinfo`.